backend/penndata/views.py

```diff
@@ -268,6 +270,130 @@ def get_queryset(self):
+                ) + datetime.timedelta(minutes=30)
+            except NonExistentTimeError:
+                # daylight savings time
+                continue
+
+            # use snapshots before and after the hour_date to interpolate
+            before = snapshots.filter(date__lte=hour_date).order_by("-date").first()
+            after = snapshots.filter(date__gte=hour_date).order_by("date").first()
+
+            before_val = getattr(before, field, None)
+            after_val = getattr(after, field, None)
+
+            if before_val is None or after_val is None:
+                usage[hour] = self.safe_add(before_val, after_val)
+                # set to None if latest entry was more than an hour ago (to avoid extrapolation)
+                if (
+                    after_val is None
+                    and before
+                    and hour_date - datetime.timedelta(hours=1) > before.date
+                ):
+                    usage[hour] = None
+            else:
+                # linear interpolation
+                usage[hour] = (
+                    (
+                        before_val
+                        + (after_val - before_val)
+                        * (hour_date - before.date).total_seconds()
+                        / (after.date - before.date).total_seconds()
+                    )
+                    if before.date != after.date
+                    else after_val
+                )
+
+        if all(x == 0 for x in usage):  # location probably closed - don't count in aggregate
+            return [None] * 24
+        return usage
+
+    def get_usage(self, room, date, num_samples, group_by, field):
+        unit = 1 if group_by == "day" else 7  # skip by 1 or 7 days
+        usage_agg = [(None, 0)] * 24  # (sum, count) for each hour
+        min_date = timezone.localtime().date()
+        max_date = date - datetime.timedelta(days=unit * (num_samples - 1))
+
+        for i in range(num_samples):
+            curr = date - datetime.timedelta(days=i * unit)
+            usage = self.get_usage_on_date(room, curr, field)  # usage for curr
+            # incoporate usage safely considering None (no data) values
+            usage_agg = [
+                (self.safe_add(acc[0], val), acc[1] + (1 if val is not None else 0))
+                for acc, val in zip(usage_agg, usage)
+            ]
+            # update min and max date if any data was logged
+            if any(usage):
+                min_date = min(min_date, curr)
+                max_date = max(max_date, curr)
+
+        ret = [x[0] / x[1] if x[0] and x[1] else 0 for x in usage_agg]
+        return ret, min_date, max_date
+
+    def get(self, request, room_id):
+        """
+        GET: returns the usage in terms of count or capacity of a fitness center for a given date
+        per hour aggregated by day or week for a given number of days
+        """
+
+        room = get_object_or_404(FitnessRoom, pk=room_id)
```

Comment:
perhaps move this line of code below all the error checking. avoids unnecessary query if response is going to fail anyways

Reply:
i was thinking to check the route parameter first, since if any error occurs the most problematic one is the resource doesn't exist?